src/experiment_prototype/experiment_slice.py

```diff
@@ -609,15 +627,97 @@ def check_clrfrqrange(cls, clrfrqrange, values):
+        # multiple of the clock divider that is possible by the USRP - this default value set
+        # here is not exact (center freq is never exactly 12 MHz).
+
+        # convert from kHz to Hz to get correct clock divider. Return the result back in kHz.
+        clock_multiples = options.usrp_master_clock_rate / 2 ** 32
+        clock_divider = math.ceil(txctrfreq * 1e3 / clock_multiples)
+        txctrfreq = (clock_divider * clock_multiples) / 1e3
+
+        return txctrfreq
+
+    @validator('rxctrfreq', always=True)
+    def check_rxctrfreq(cls, rxctrfreq):
+        if not rxctrfreq:
+            rxctrfreq = 12000.0  # Default value when not set
+
+        # Note - rxctrfreq set here and modify the actual center frequency to a
+        # multiple of the clock divider that is possible by the USRP - this default value set
+        # here is not exact (center freq is never exactly 12 MHz).
+
+        # convert from kHz to Hz to get correct clock divider. Return the result back in kHz.
+        clock_multiples = options.usrp_master_clock_rate / 2 ** 32
+        clock_divider = math.ceil(rxctrfreq * 1e3 / clock_multiples)
+        rxctrfreq = (clock_divider * clock_multiples) / 1e3
+
+        return rxctrfreq
+
+    @validator('tx_maxfreq', always=True)
+    def check_tx_maxfreq(cls, tx_maxfreq, values):
+        tx_maxfreq = options.max_freq
+        
+        if 'txctrfreq' in values:
+            new_max = values['txctrfreq'] * 1000 + (values['tx_bandwidth'] / 2.0) - values['transition_bandwidth']
+            if new_max < options.max_freq:
+                tx_maxfreq = new_max
+
+        return tx_maxfreq
+
+    @validator('tx_minfreq', always=True)
+    def check_tx_minfreq(cls, tx_minfreq, values):
+        tx_minfreq = options.min_freq
+
+        if 'txctrfreq' in values:
+            new_min = values['txctrfreq'] * 1000 - (values['tx_bandwidth'] / 2.0) + values['transition_bandwidth']
+            if new_min > options.min_freq:
+                tx_minfreq = new_min
+
+        return tx_minfreq
+
+    @validator('rx_maxfreq', always=True)
+    def check_rx_maxfreq(cls, rx_maxfreq, values):
+        rx_maxfreq = options.max_freq
+
+        if 'rxctrfreq' in values:
+            new_max = values['rxctrfreq'] * 1000 + (values['rx_bandwidth'] / 2.0) - values['transition_bandwidth']
+            if new_max < options.max_freq:
+                rx_maxfreq = new_max
+
+        return rx_maxfreq
+
+    @validator('rx_minfreq', always=True)
+    def check_rx_minfreq(cls, rx_minfreq, values):
+        rx_minfreq = options.min_freq
+
+        if 'rxctrfreq' in values:
+            new_min = values['rxctrfreq'] * 1000 - (values['rx_bandwidth'] / 2.0) + values['transition_bandwidth']
+            if new_min > options.min_freq:
+                rx_minfreq = new_min
+
+        return rx_minfreq
+
     @validator('freq')
-    def check_freq(cls, freq):
+    def check_freq(cls, freq, values):
         if not freq:
             return
```

Comment:
Linked to comment above, we should make `freq` a required field, then this could be removed.